### PR TITLE
Synchrona rev b leds service

### PR DESCRIPTION
**Pull request by tachicialex**
c5f0b8f9c05b - synchrona: move LEDs control to separate service
f1ffa358a033 - synchrona: add direction to script

Tested with the latest Kuiper Image + newest rpi-ad9545-hmc7044 /kernel image/modules.

### CODE REVIEW

File: installers/ad-synchrona-14-leds.sh [hunk 1]
```diff
@@ -55,6 +55,9 @@ do
 		LED_STATUS=$STATUS_RED
 	fi
 
+	echo "out" > "/sys/class/gpio/gpio$RED_LED1/direction"
+	echo "out" > "/sys/class/gpio/gpio$GREEN_LED1/direction"
```
Comment (nunojsa) [inline installers/ad-synchrona-14-leds.sh:70]: Yeah, I saw a similar thing on the main service. While in there it's more odd to me what could have been the issue/race, in here I can see what could be the race... When we reload (remove + aload) the overlay and the ad9545 device is unbind, the pinctrl settings are also removed which brings the pins back to their default (which might be input). If you happen to try to set a value at that moment, the issue pops up...

So yeah, for now let's have this workaround and keep things simple but I'm not sure if we would need some kind of synchronization between this script and the reload_dtb.sh

File: installers/common.sh [hunk 2]
```diff
@@ -204,20 +206,37 @@ function _setup_synchrona_service() {
...
+    StartLimitIntervalSec=0
+
+    [Service]
+    Type=simple
+    User=root
+    ExecStartPre=-/usr/bin/sh -c "/usr/bin/echo 12 > /sys/class/gpio/export"
+    ExecStartPre=-/usr/bin/sh -c "/usr/bin/echo 16 > /sys/class/gpio/export"
+    ExecStart=/usr/bin/bash /etc/raspap/synchrona/ad-synchrona-14-leds.sh > /dev/null 2>&1 &
     ExecStopPost=-/usr/bin/sh -c "/usr/bin/echo 12 > /sys/class/gpio/unexport
     ExecStopPost=-/usr/bin/sh -c "/usr/bin/echo 16 > /sys/class/gpio/unexport
```
Comment (nunojsa) [inline installers/common.sh]: This is missing `"` in the end. In fact, this is a bug introduced when the led script was added (which I completely failed to see during review). So, the fix for this should be in it's own patch before doing moving this to it's own service...

File: installers/common.sh [hunk 3]
```diff
@@ -204,20 +206,37 @@ function _setup_synchrona_service() {
...
+    Description=Synchrona LEDs bash service
+    After=synchrona.service
+    PartOf=synchrona.service
+    StartLimitIntervalSec=0
+
+    [Service]
+    Type=simple
+    User=root
+    ExecStartPre=-/usr/bin/sh -c "/usr/bin/echo 12 > /sys/class/gpio/export"
+    ExecStartPre=-/usr/bin/sh -c "/usr/bin/echo 16 > /sys/class/gpio/export"
```
Comment (nunojsa) [inline installers/common.sh]: nit: Can't we do this in the script? So we check if the gpio is already exported and if not, we export them. The same goes for the unexport part that we can do when leaving the script (you might need to trap SIGSTOP in order to terminate the main loop). In this way we could leave the service file a bit more simple...

File: installers/common.sh [hunk 4]
```diff
@@ -204,20 +206,37 @@ function _setup_synchrona_service() {
...
+    ExecStart=/usr/bin/bash /etc/raspap/synchrona/ad-synchrona-14-leds.sh > /dev/null 2>&1 &
```
Comment (nunojsa) [inline installers/common.sh]: Do we still need to run the script in the background? Im fairly sure that systemd will already "daemonize" the script... I would also not redirect all the log to /dev/null as it might be useful to run `journalctl -u synchrona_leds` to look for possible issues

**Review by nunojsa — COMMENTED** [review]
Just some remarks from me...

**Comment by tachicialex** [comment]
e2a4bad9465d - installers: common: Removed pre/post from LEDs service
cb3e656155df - installers: common: move LEDs control to separate service
71d198cedb1c - installers: common: fix missing quotes
d962b41c123d - installers: ad-synchrona-14-leds: add direction to script

Changelog v1 - > v2:
- fixed missing " at the end of unexport
- ad-synchrona-14-leds.sh don't run script in background
- no longer redirect ad-synchrona-14-leds.sh output to /dev/null, for visibility in logs
- moved export/unexport from service file to ad-synchrona-14-leds.sh

Tested with latest kuiper + dt

File: installers/ad-synchrona-14-leds.sh [hunk 5]
```diff
@@ -17,6 +17,24 @@ AD9545_LOCKED=0
 HMC7044_LOCKED=0
 AD9545_FREERUN=0
 
+trap synchrona_leds_exit SIGTERM
+function synchrona_leds_exit()
```
Comment (nunojsa) [inline installers/ad-synchrona-14-leds.sh]: nit: typically in bash we don't really use `function`. Just `synchrona_leds_exit() {`

File: installers/ad-synchrona-14-leds.sh [hunk 6]
```diff
@@ -17,6 +17,24 @@ AD9545_LOCKED=0
...
+{
+	echo $RED_LED1 > "/sys/class/gpio/unexport"
+	echo $GREEN_LED1 > "/sys/class/gpio/unexport"
+	exit
```
Comment (nunojsa) [inline installers/ad-synchrona-14-leds.sh]: I was more thinking in having a control variable to break the loop. But yeah, likely is not a big deal but feels wrong to exit a program in a signal handler (even though it's a script)

File: installers/ad-synchrona-14-leds.sh [hunk 7]
```diff
@@ -17,6 +17,24 @@ AD9545_LOCKED=0
...
+}
+
+if [ ! -d "/sys/class/gpio/gpio$RED_LED1" ]
+then
+	echo $RED_LED1 > "/sys/class/gpio/export"
+fi
```
Comment (nunojsa) [inline installers/ad-synchrona-14-leds.sh]: nit: this could be a one liner `[ ! -d "/sys/class/gpio/gpio$RED_LED1" ] && echo $RED_LED1 > "/sys/class/gpio/export"`

**Review by nunojsa — APPROVED** [review]
Just some minor nits that you can ignore if you want... Anyways, you need to rebase and surely test it :)

**Comment by tachicialex** [comment]
Changelog v2 -> v3:
- rebased over synchrona_revB
- removed "function"
- moved all "if -d" gpio dir checking to oneliners
- added a control var. to exit while() loop nicer

Pushed here just for testing. Not tested yet.

**Review by nunojsa — APPROVED** [review]
Let me know when it's tested...

**Comment by tachicialex** [comment]
Ok, script is working well, when service gets stopped, LED turns red and script process gets killed. Next start ,script process is back running and changing the LED color accordingly. Tested today.
